File: lab10-iterative_linear/lab10-iterative_linear_f90/jacobi_solve_mpi.f90
! -*- Mode: Fortran90; -*-
!-----------------------------------------------------------------
! Daniel R. Reynolds
! SMU Mathematics
! Math 4370 / 6370
!=================================================================

subroutine jacobi_solve(a, b, c, u, r, res, local_N, delta, &
     maxiter, iters, comm, ierr)
  !-----------------------------------------------------------------
  ! Description:
  !   Solves the linear system
  !           a(k)*u(k-1) + b(k)*u(k) + c(k)*u(k+1) = r(k),
  !   using a parallelized Jacobi iterative solver.
  !-----------------------------------------------------------------
  !======= Inclusions ===========
  use mpi

  !======= Declarations =========
  implicit none

  integer, intent(in) :: local_N, comm, maxiter
  integer, intent(out) :: ierr
  double precision, dimension(local_N) :: a, b, c, u, r, res
  double precision, intent(in) :: delta
  integer, intent(out) :: iters
  double precision :: resid2
  integer :: k

  !======= Internals ============

  ! compute initial linear residual
  call linresid(a, b, c, u, r, res, resid2, local_N, comm, ierr)
  if (ierr /= 0) then
     write(0,*) ' jacobi_solve error: linresid failed!'
     return
  end if

  ! iterate until resid2 < delta (or maxiter iterations)
  do iters=0,maxiter

     ! check for convergence
     if (resid2 < delta)  exit

     ! update u = u - diag(b)^{-1}*res
     u = u - res/b

     ! compute linear residual
     call linresid(a, b, c, u, r, res, resid2, local_N, comm, ierr)
     if (ierr /= 0) then
        write(0,*) ' jacobi_solve error: linresid failed!'
        return
     end if
  end do

  if (iters == maxiter) then
     write(0,*) '  jacobi_solve warning: reached maximum iteration limit!'
  end if

  return

end subroutine jacobi_solve
!=================================================================

File: lab10-iterative_linear/lab10-iterative_linear_cxx/jacobi_solve_mpi.cpp
/* Daniel R. Reynolds
   SMU Mathematics
   Math 4370 / 6370 */

// Inclusions
#include <stdlib.h>
#include <iostream>
#include "mpi.h"

// Prototypes
int linresid(double *, double *, double *, double *, double *,
	     double *, double &, int, MPI_Comm);


/* Description: Solves the linear system
             a(k)*u(k-1) + b(k)*u(k) + c(k)*u(k+1) = r(k),
     using a parallelized Jacobi iterative solver. */
int jacobi_solve(double *a, double *b, double *c, double *u,
		 double *r, double *res, int loc_N, double delta,
		 int maxiter, int &iters, MPI_Comm comm) {

  // compute initial linear residual
  double resid2;
  if (linresid(a, b, c, u, r, res, resid2, loc_N, comm) != 0) {
    std::cerr << " jacobi_solve error: linresid failed!\n";
    return 1;
  }

  // iterate until resid2 < delta (or maxiter iterations)
  for (iters=0; iters<=maxiter; iters++) {

    // check for convergence
    if (resid2 < delta)  break;

    // update u = u - diag(b)^{-1}*res
    for (int i=0; i<loc_N; i++)   u[i] = u[i] - res[i]/b[i];

    // compute linear residual
    if ( linresid(a, b, c, u, r, res, resid2, loc_N, comm) != 0) {
      std::cerr << " jacobi_solve error: linresid failed!\n";
      return 1;
    }
  } // for iters

  if (iters == maxiter)
    std::cerr << " jacobi_solve warning: reached maximum iteration limit!\n";

  return 0;

} // end jacobi_solve
